railties/lib/rails/commands/restart/restart_command.rb

```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+
+module Rails
+  module Command
+    class RestartCommand < Base # :nodoc:
+      desc "restart", "Restart app by touching tmp/restart.txt"
+      def perform
+        require "fileutils"
+
+        FileUtils.mkdir_p "tmp"
+        FileUtils.touch   "tmp/restart.txt"
```

Comment:
I was curious what would happen if this command was run from outside of a Rails application directory.

On main it errors:
```
$ ./min/bin/rails restart
rails aborted!
No Rakefile found (looking for: rakefile, Rakefile, rakefile.rb, Rakefile.rb)
/home/hartley/src/github.com/skipkayhil/rails/railties/lib/rails/commands/rake/rake_command.rb:20:in `block in perform'
/home/hartley/src/github.com/skipkayhil/rails/railties/lib/rails/commands/rake/rake_command.rb:18:in `perform'
/home/hartley/src/github.com/skipkayhil/rails/railties/lib/rails/command.rb:152:in `invoke_rake'
/home/hartley/src/github.com/skipkayhil/rails/railties/lib/rails/command.rb:73:in `block in invoke'
/home/hartley/src/github.com/skipkayhil/rails/railties/lib/rails/command.rb:145:in `with_argv'
/home/hartley/src/github.com/skipkayhil/rails/railties/lib/rails/command.rb:69:in `invoke'
/home/hartley/src/github.com/skipkayhil/rails/railties/lib/rails/commands.rb:18:in `<top (required)>'
min/bin/rails:4:in `require'
min/bin/rails:4:in `<main>'
(See full trace by running task with --trace)
```

but with this branch it will create `tmp/restart.txt` in whatever the current directory is. I think we need to use `Rails.root.join` here?

Reply:
@skipkayhil  I tried this locally.
Running the task from the `#{Rails.root}/config` directory still updated `#{Rails.root}/tmp/restart.txt`.
I think we can only run `new` and `plugin` outside a Rails directory:
https://github.com/rails/rails/blob/2eed4dc0afd1b82b2c12c6f77ab7271e72699168/railties/lib/rails/cli.rb#L12-L22

Reply:
I think what @skipkayhil meant is call `bin/rails` (not gem `rails`) from outside the app directory.  For example, see https://community.fly.io/t/no-rakefile-found-upon-running-example-command/6962.

With `main`, doing so will raise "No Rakefile found", but with this PR it will silently fail.

Personally, I think it would be nice to support the use case in that thread.

Reply:
> call `bin/rails` (not gem `rails`) from outside the app directory.

Exactly. If I understand correctly, being able to do this is intentionally supported by setting `BUNDLE_GEMFILE` in `config/boot`

https://github.com/rails/rails/blob/3025eaa09154f5f031f5c2e0e5b59553ff5c902b/railties/lib/rails/generators/rails/app/templates/config/boot.rb.tt#L1

> With `main`, doing so will raise "No Rakefile found", but with this PR it will silently fail.

This is the point I think is important. Before, calling `bin/rails` for this particular command wasn't possible because it was a Rake task. However, since we're now enabling people to call this command from outside of a Rails project I think it's important we aren't changing files outside of the project.

With this command, "silently failing" means a folder and file are created outside of the project directory. I think this is even more important for `log:clear` since "silently failing" means _removing files_ outside of the project directory.

Reply:
Thanks for the explanation @skipkayhil ! That makes sense.
I was a bit hesitant in loading the whole application to get Rails.root for restarting, but I found there also is a `Rails::Command.root`:
https://github.com/rails/rails/blob/main/railties/lib/rails/command.rb#L107-L114

The following now works for me:
```bash
BUNDLE_GEMFILE=demo/Gemfile demo/bin/rails restart
```